nipst/post.go

```diff
@@ -3,96 +3,130 @@ package nipst
 import (
 	"fmt"
 	"github.com/spacemeshos/go-spacemesh/types"
-	"github.com/spacemeshos/merkle-tree"
 	"github.com/spacemeshos/post/config"
 	"github.com/spacemeshos/post/initialization"
 	"github.com/spacemeshos/post/proving"
 	"github.com/spacemeshos/post/shared"
 	"github.com/spacemeshos/post/validation"
-	"time"
+	"sync"
 )
 
 func DefaultConfig() config.Config {
-	cfg := config.DefaultConfig()
-	cfg.Difficulty = 5
-	cfg.NumProvenLabels = 10
-	cfg.SpacePerUnit = 1 << 10 // 1KB.
-	cfg.FileSize = 1 << 10     // 1KB.
-	return *cfg
+	return *config.DefaultConfig()
 }
 
-func verifyPost(proof *types.PostProof, space uint64, numProvenLabels uint, difficulty uint) (bool, error) {
+func verifyPost(proof *types.PostProof, space uint64, numProvenLabels uint, difficulty uint) error {
 	if space%config.LabelGroupSize != 0 {
-		return false, fmt.Errorf("space (%d) is not a multiple of LabelGroupSize (%d)", space, config.LabelGroupSize)
+		return fmt.Errorf("space (%d) is not a multiple of LabelGroupSize (%d)", space, config.LabelGroupSize)
 	}
 
 	cfg := config.Config{SpacePerUnit: space, NumProvenLabels: uint(numProvenLabels), Difficulty: uint(difficulty)}
 	validator := validation.NewValidator(&cfg)
-	err := validator.Validate((*proving.Proof)(proof))
-	if err != nil {
-		return false, err
+	if err := validator.Validate((*proving.Proof)(proof)); err != nil {
+		return err
 	}
 
-	return true, nil
+	return nil
 }
 
 type PostClient struct {
+	id          []byte
 	cfg         *config.Config
 	initializer *initialization.Initializer
+	prover      *proving.Prover
 	logger      shared.Logger
+
+	sync.RWMutex
 }
 
 // A compile time check to ensure that PostClient fully implements PostProverClient.
 var _ PostProverClient = (*PostClient)(nil)
 
-func NewPostClient(cfg *config.Config) *PostClient {
-	return &PostClient{cfg, nil, shared.DisabledLogger{}}
+func NewPostClient(cfg *config.Config, id []byte) *PostClient {
+	return &PostClient{
+		id:          id,
+		cfg:         cfg,
+		initializer: initialization.NewInitializer(cfg, id),
+		prover:      proving.NewProver(cfg, id),
+		logger:      shared.DisabledLogger{},
+	}
 }
 
-func (c *PostClient) initialize(id []byte, timeout time.Duration) (commitment *types.PostProof, err error) {
-	// TODO(moshababo): implement timeout
-	//TODO: implement persistence
-	if c.cfg.SpacePerUnit%merkle.NodeSize != 0 {
-		return nil, fmt.Errorf("space (%d) is not a multiple of merkle.NodeSize (%d)", c.cfg.SpacePerUnit, merkle.NodeSize)
-	}
-	c.initializer = initialization.NewInitializer(c.cfg, id)
+func (c *PostClient) Initialize() (commitment *types.PostProof, err error) {
```

Comment:
`Initialize` is a very long operation, are you sure that you have to hold the lock for the entire duration? For example, if, during initialization, the wallet will call `StartPost` it will be delayed until the previous `Initialize` call returns.

Reply:
Good point, although that shouldn't happen, because the atx builder state protects against multiple calls.
I should anyway reduce  the lock to protect only the struct state, as originally intended. To do so, I can change it to from `Lock` to `RLock`. 
But `SetParams` must remain `Lock`, and so the problem you described can still occur, if not protected in a higher scope. 
I can use atomic ops and variable state instead, and return errors instead of introducing waits, but it looks like an overkill to me.

Let me know what you think.